cardano-node-socket-emulator/src/Cardano/Node/Socket/Emulator/Server.hs

```diff
@@ -601,22 +606,37 @@ txSubmissionServer
   -> TxSubmission.LocalTxSubmissionServer (Shelley.GenTx block) (ApplyTxErr block) IO ()
 txSubmissionServer state =
   TxSubmission.LocalTxSubmissionServer
-    { TxSubmission.recvMsgSubmitTx =
-        \tx -> do
-          case tx of
-            (Consensus.HardForkGenTx (Consensus.OneEraGenTx (S (S (S (S (S (Z tx')))))))) -> do
-              let Shelley.ShelleyTx _txid shelleyEraTx = tx'
-              let ctx = CardanoEmulatorEraTx (C.ShelleyTx C.ShelleyBasedEraBabbage shelleyEraTx)
-              _ <-
-                modifyMVar_
-                  state
-                  (pure . over (socketEmulatorState . emulatorState . E.esChainState) (Chain.addTxToPool ctx))
-              pure ()
-            _ -> pure ()
-          return (TxSubmission.SubmitSuccess, txSubmissionServer state)
+    { TxSubmission.recvMsgSubmitTx = \tx -> (,txSubmissionServer state) <$> submitTx state tx
     , TxSubmission.recvMsgDone = ()
     }
 
+submitTx
+  :: (block ~ CardanoBlock StandardCrypto)
+  => MVar AppState
+  -> Shelley.GenTx block
+  -> IO (TxSubmission.SubmitResult (ApplyTxErr block))
+submitTx state tx = case C.fromConsensusGenTx tx of
+  C.TxInMode C.ShelleyBasedEraBabbage shelleyTx -> do
+    AppState
+      (SocketEmulatorState (E.EmulatorState (Chain.ChainState _ _ index slot _) _ _) _ _)
+      _
+      params <-
+      readMVar state
+    case Validation.validateAndApplyTx params (fromIntegral slot) (fromPlutusIndex index) shelleyTx of
+      Left err ->
+        pure $
+          TxSubmission.SubmitFail
+            ( Consensus.HardForkApplyTxErrFromEra
+                (Consensus.OneEraApplyTxErr (S (S (S (S (S (Z (WrapApplyTxErr err))))))))
+            )
+      Right _ -> do
+        let ctx = CardanoEmulatorEraTx shelleyTx
+        modifyMVar_
+          state
+          (pure . over (socketEmulatorState . emulatorState . E.esChainState) (Chain.addTxToPool ctx))
+        pure TxSubmission.SubmitSuccess
+  _ -> pure $ TxSubmission.SubmitSuccess -- should be SubmitFail HardForkApplyTxErrWrongEra, but the Mismatch type is complicated
```

Comment:
wouldn't it be better to still fail here with some arbitrary Mismatch Text for the two eras?

Reply:
As far as I can tell that's not possible.